### PR TITLE
Add immutable, functional linked list in JavaScript

### PR DESCRIPTION
**Pull request by Bertrand31**
The methods are inspired by Scala's Standard Library

### CODE REVIEW

File: javascript/data_structures/singly_linked_list.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  flatten = (tabbed = null) => {
+    if (this.next === null && tabbed === null) return this;
+    if (this.next === null) return tabbed.flatten();
+    if (this.value instanceof LinkedList) return this.value.flatten(this.next);
+    return new LinkedList(this.value, this.next.flatten(tabbed));
+  }
+
+  flatMap = fn => this.map(fn).flatten()
+
+  take = (until, hops = 1) => {
```
Comment (nicholette) [inline javascript/data_structures/singly_linked_list.js]: @Bertrand31 What is this function supposed to do? 🤔

**Comment by Bertrand31** [comment]
By the way, this linked list is implemented as a _functor_ (it has a `map`) and even a _monad_ (it has both `map` and `flatten`).

**Comment by Bertrand31** [comment]
This hasn't really been optimized for performance. But how would you suggest to improve these? @moroine (bear in mind this is an immutable data structure).

**Comment by moroine** [comment]
@Bertrand31 I don't really know. Also would be nice to have a `fromArray`

**Comment by Bertrand31** [comment]
There you go @moroine

**Pull request merged**